ds0_updater.py

```diff
@@ -5,13 +5,36 @@
 import logger
 import conf
 
+# keep it super simple for now because we don't really implement --allowed-keys yet.
+def verify_latest_commit():
+    allowed_keys = ds0_utils.allowed_keys()
+
+    # ensure the latest commit is signed and verified
+    while True:
+        for key in allowed_keys:
+            logger.exe('gpg --import ' + key['file'], expectError=True)
+        output = logger.exe('git log --pretty="format:%G?" --show-signature HEAD^..HEAD')
+
+        if "Can't check signature" in output[0]:
+            logger.info('gpg keys cleared on startup. Trying again...')
+            continue
+
+        verified = False
+        for key in allowed_keys:
+            if key['rsa_check'] in output[0] and key['signature_check'] in output[0]:
+                verified = True
+        if not verified:
+            logger.error('Scripts using a non-signed commit. Please ensure commit is valid.')
+            logger.error('    If it was a missed signature, feel free to open a ticket at https://github.com/riptano/ComboAMI.')
+        break
+
 # Update the AMI codebase if it's its first boot
 if not conf.get_config("AMI", "CompletedFirstBoot"):
     # check if a specific commit was requested
     force_commit = ds0_utils.required_commit()
 
     # update the repo
-    logger.exe('git pull')
+    logger.exe('git pull ' + ds0_utils.repository())
```

Comment:
What I was getting at with leaking the remote ("origin") into the second-half of the --repository flag is this:

```
% git remote # No remote repositories defined

% git branch -a # No branches from any remotes checked out locally
* scratch-branch

% git remote add origin git@github.com:riptano/ComboAMI.git                                                
% git fetch                                                                                                !10195
From github.com:riptano/ComboAMI
 * [new branch]      2.0        -> origin/2.0
 * [new branch]      2.1        -> origin/2.1
 * [new branch]      2.2        -> origin/2.2
 * [new branch]      2.3        -> origin/2.3
 * [new branch]      2.4        -> origin/2.4
 * [new branch]      2.5        -> origin/2.5
 * [new branch]      dev        -> origin/dev
 * [new branch]      dev-2.5    -> origin/dev-2.5

% git reset --hard 2.5 # Resetting to a bare-branch name doesn't work
fatal: ambiguous argument '2.5': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'

% git reset --hard origin/2.5 # But resetting to a remote/branch combo does
HEAD is now at e5e3d41 Merge pull request #71 from riptano/dev-2.5

% git reset --hard 33f5011 # As does resetting to a commit-hash
HEAD is now at 33f5011 Reorder chowing
```

A few other useful commit-reference formats can be passed directly to git-reset as well. So without any custom-code to detect or lookup branch-names vs commit-hashes, both work in the above scheme (with the drawback that branch-names must be prefixed by a remote-name). You may be able to improve upon this with some combination of checkout and pull but this is quite simple to implement and does quite a lot.

At this point I'd say use your judgement, you've got a firm understanding of the issues in play and given the code and analysis you've submitted so far I'm sure whatever you come up with will be mergeable.